### PR TITLE
BUG: add attribute '__array__' for mars dataframe

### LINKED ISSUE #248: [BUG] AttributeError: 'series' object has no attribute 'array' when calling np.array(df)
### Describe the bug
When calling `np.array(df)` in my code, I encountered the following error: "`AttributeError: 'series' object has no attribute '__array__'`".

### To Reproduce
To help us to reproduce this bug, please provide information below:

1. Your Python version: python 3.8
2. The version of Xorbits you use: the latest, 0.2.0
3. Versions of crucial packages:
numpy==1.22.4
5. Minimized code to reproduce the error.

```
import numpy as np
import xorbits.pandas as pd

# code that triggers the error
data = np.array(df)
```

### Expected behavior
I expected the `np.array()` function to create a numpy array from the Pandas DataFrame `df`.

### Additional context
N/A

### PR DESCRIPTION
**Pull request by Bojun-Feng**
## What do these changes do?

- Fix error when mixing original numpy and xorbits.pandas dataframe
- Add tests in pandas module testing conversions between dataframe and numpy array
- Update potentially misleading error message

The original error message `AttributeError: 'series' object has no attribute '__array__'` seems to be incorrect, as the dataframe clearly has a callable attribute named `__array__` but is unable to run it:

<p align="center">
<kbd>

</kbd>
</p>

After poking around in the source code, it seemed that the following three cases hold at the same time:
1. xorbits dataframe does have the feature
2. xorbits dataframe is a wrapper of the mars dataframe
3. mars dataframe does not have the feature

<p align="center">
<kbd>

</kbd>
</p>

The `__array__` feature was added to mars dataframes, and the error message was also updated to pinpoint feature-related errors to mars objects instead of to xorbits objects.

## Related issue number

Fixes #248

## Check code requirements

- [x] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

**Comment by UranusSeven** [comment]
Hi Feng, thank you for your contribution. For lint problems, please refer this document: https://doc.xorbits.io/en/latest/development/contributing.html#autofixing-formatting-errors

you can setup the pre-commit env by running `pre-commit install` under the project root dir.

**Comment by UranusSeven** [comment]
If you don't have pre-commit installed, you could run `pip install pre-commit`.

**Comment by Bojun-Feng** [comment]
> If you don't have pre-commit installed, you could run `pip install pre-commit`.

That is really helpful, thank you so much! I am planning to write up some tests as well to test the compatibility between np, pd, xnp, and xpd. However, I'm not sure whether I should put it in the tests for numpy or pandas. Which one do you think fits better? Any feedback would be much appreciated!

**Comment by Bojun-Feng** [comment]
Quick Update: 
- New test cases have been added in pandas/tests/test_compat.py
- Several format-only commits were merged for brevity
- Pre-commit was run locally to edit code formatting

**Comment by Bojun-Feng** [comment]
The pre-commit is behaving weirdly in my local repository, as prettier constantly tries to read the _mars directory as a file and gives the EISDIR error. I am currently looking up the issue. If anyone else is having the same problem please let me know.

File: python/xorbits/web/ui/resources/favicon.svg
Comment (qianduoduo0904) [inline python/xorbits/web/ui/resources/favicon.svg:1]: A bit weird to see this file modified, could you revert this one?

**Comment by UranusSeven** [comment]
> 

Could you post your command and the detailed error?

pre-commit is configured to ignore the _mars dir, but if you run `pre-commit run --all-files`, which you shouldn't normally do, the _mars dir will be read.

File: python/xorbits/_mars/_version.py [hunk 1]
```diff
@@ -10,12 +10,12 @@
 """Git implementation of _version.py."""
 
 import errno
+import functools
```
Comment (UranusSeven) [inline python/xorbits/_mars/_version.py:13]: Please do not include unrelated changes into this PR. Changes that fix the order of import statements should be an individual PR.

File: python/xorbits/_mars/dataframe/initializer.py [hunk 2]
```diff
@@ -120,6 +121,11 @@ def __init__(
             df = df.rebalance(num_partitions=num_partitions)
         super().__init__(df.data)
 
+    # Attribute for numpy array conversion, defined in pandas/pandas/core/generic.py
```
[the remark this answers is not in the document]
Comment (UranusSeven) [reply python/xorbits/_mars/dataframe/initializer.py:124]: Also the implementation of __array__ is actually not that simple. You need to fetch all the distributed data chunk. Have a look at `xorbits._mars.dtaframe.core.DataFrame.__str__` may be helpful.

[the remark this answers is not in the document]
Comment (Bojun-Feng) [reply python/xorbits/_mars/dataframe/initializer.py:124]: I looked through the codebase, it seems that the `DataFrame` class has no `.__str__()` method, instead, the method is defined on `DataFrameData`.

I was just wondering where should I add the attribute. The initializer inherits from `DataFrame`, so it seems to make sense to add attributes to the `DataFrame` class. However, it seems that the actual data processing all happens in `DataFrameData` class.

My biggest confusion is that `DataFrame` does not seem to inherit anything from `DataFrameData` (`DataFrame` -> `HasShapeTileable` -> `Tileable` -> `Entity` -> `Serializable`), and neither does `DataFrameData` inherit from `DataFrame`, so I am unsure what is the relationship between them and which class does Xorbits use.

Sorry for the late reply, any advice would be much appreciated.

**Comment by Bojun-Feng** [comment]
Quick Update:

- Unnecessary formatting changes have been removed
- Tests have been re-written in a more readable way using pytest parametrize
- The code is working fine on my local computer when there are previous sessions to `fetch()` from

Problem:

When there are no sessions to `fetch()`, the current code transfers everything into a mars tensor and then into a numpy array. However, I am running into issues where **the mars tensor keeps trying to assign strings into floats in its `__array__` attribute, which leads to an error** when there is a string in the dataframe. Right now I am thinking about changing the `__array__` attribute of mars tensors so it assigns strings to strings. 

However, I am still not familiar with the code base, so I have these two questions:
- Is there an intentional reason why tensors don't work well with strings? Or is this just a bug?
- Is translating everything into a tensor a good strategy? Are there any better functions/classes I can look into?

Any advice would be much appreciated!

**Comment by UranusSeven** [comment]
> 
You are right, the implementation of `DataFrame.__mars_tensor__` contains a bug. I'm able to reproduce it and have submitted an issue to track the bug #339.

You can use `DataFrame.to_tensor` instead. It convert the dataframe into a tensor whose dtype is inferred from the dataframe's dtypes.

File: python/xorbits/core/data.py [hunk 3]
```diff
@@ -272,6 +278,15 @@ def __repr__(self):
             run(self)
             return self.data.__repr__()
 
+    def __array__(self):
+        from .execution import run
+
+        if self._own_data():
+            return self.data._mars_entity.op.data.__array__()
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:285]: Better to add a test to cover this line.

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Review by ChengjieLi28 — APPROVED** [review]
